Source/Dafny/Plugins/Rewriter.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System.Diagnostics.Contracts;
+
+namespace Microsoft.Dafny.Plugins {
+  /// <summary>
+  /// A class that plugins should extend, in order to provide an extra Rewriter to the pipeline.
+  ///
+  /// If the plugin defines no Configuration, then Dafny will instantiate every sub-class
+  /// of Rewriter from the plugin, providing them with an ErrorReporter as the first and only argument.
+  /// </summary>
+  public abstract class Rewriter {
```

Comment:
Thanks for the effort in making the contract as clear as possible here. I think this meets the bar for the feature as long as its marked as experimental. I do think we'll need more of a mechanism to break the build if we change the various AST types in a way that break plugins before we fully commit, though, so good to start thinking about that now.

Reply:
For that, we need access relevant plugins that we intend to support, indeed. This will be the case for compilers soon.

Reply:
No, that's not the right model. We need to explicitly decide on exactly what parts of the implementation plugins have access to, and at the very least document it clearly. Even better, we should use mechanisms to make sure we don't expose things we don't intend to keep the same in the future (i.e. `public` vs `internal`), and don't accidentally change things we said we wouldn't (i.e. API compatibility tests, which there may be tooling out there for on .NET).

Even with all that, there will still be unfortunate cases where users start relying on something you didn't intend them to (I refer you to the infamous XKCD on that subject: https://xkcd.com/1172/ :) ) and changing things without breaking clients will sometimes be tricky. If you make your API "whatever the users we're aware of are using" you're making it that much harder to scale your resources and far more likely you're going to break people and/or tying your hands for refactoring internals for the better in the future.

Reply:
</rant> :)

Note I'm not asking for any deeper changes on this PR right now - it will be great to have folks experiment as long as they know the API isn't finalized yet!

Reply:
I totally agree on that !